### PR TITLE
std: introduce GeneralPurposeAllocator

### PR DESCRIPTION
**Pull request by andrewrk**
`std.GeneralPurposeAllocator` is now available. It is a function that
takes a configuration struct (with default field values) and returns an
allocator. There is a detailed description of this allocator in the
doc comments at the top of the new file.

The main feature of this allocator is that it is *safe*. It
prevents double-free, use-after-free, and detects leaks.

Some deprecation compile errors are removed.

The Allocator interface gains `old_align` as a new parameter to
`resizeFn`. This is useful to quickly look up allocations.

`std.heap.page_allocator` is improved to use mmap address hints to avoid
obtaining the same virtual address pages when unmapping and mapping
pages. The new general purpose allocator uses the page allocator as its
backing allocator by default.

`std.testing.allocator` is replaced with usage of this new allocator,
which does leak checking, and so the LeakCheckAllocator is retired.

stage1 is improved so that the `@typeInfo` of a pointer has a lazy value
for the alignment of the child type, to avoid false dependency loops
when dealing with pointers to async function frames.

The `std.mem.Allocator` interface is refactored to be in its own file.

`std.Mutex` now exposes the dummy mutex with `std.mutex.Dummy`.

This allocator is great for debug mode, however it needs some work to
have better performance in release modes. The next step will be setting
up a series of tests in ziglang/gotta-go-fast and then making
improvements to the implementation.

### CODE REVIEW

File: lib/std/heap/general_purpose_allocator.zig [hunk 1]
```diff
@@ -0,0 +1,927 @@
...
+
+        const mutex_init = if (config.thread_safe) std.Mutex{} else std.mutex.Dummy{};
+
+        const stack_n = config.stack_trace_frames;
+        const one_trace_size = @sizeOf(usize) * stack_n;
+        const traces_per_slot = 2;
+
+        pub const Error = std.mem.Allocator.Error;
+
+        const small_bucket_count = math.log2(page_size);
```
Comment (daurnimator) [inline lib/std/heap/general_purpose_allocator.zig:151]: This should be configurable.

File: lib/std/heap/general_purpose_allocator.zig [hunk 2]
```diff
@@ -0,0 +1,927 @@
...
+        }
+    };
+}
+
+const TraceKind = enum {
+    alloc,
+    free,
+};
+
+fn up_to_nearest_power_of_2(comptime T: type, n: T) T {
```
Comment (daurnimator) [inline lib/std/heap/general_purpose_allocator.zig:588]: Use `math.ceilPowerOfTwo`

File: lib/std/heap/general_purpose_allocator.zig [hunk 3]
```diff
@@ -0,0 +1,927 @@
...
+        };
+        const LargeAllocTable = std.HashMapUnmanaged(usize, LargeAlloc, hash_addr, eql_addr, false);
+
+        // Bucket: In memory, in order:
+        // * BucketHeader
+        // * bucket_used_bits: [N]u8, // 1 bit for every slot; 1 byte for every 8 slots
+        // * stack_trace_addresses: [N]usize, // traces_per_slot for every allocation
+
+        const BucketHeader = struct {
+            prev: *BucketHeader,
```
Comment (daurnimator) [inline lib/std/heap/general_purpose_allocator.zig:178]: Shouldn't this be a `std.LinkedList`?

File: lib/std/heap/general_purpose_allocator.zig [hunk 4]
```diff
@@ -0,0 +1,920 @@
...
+};
+
+fn up_to_nearest_power_of_2(comptime T: type, n: T) T {
+    var power: T = 1;
+    while (power < n)
+        power *= 2;
+    return power;
+}
+
+fn hash_addr(addr: usize) u32 {
```
Comment (Sahnvour) [inline lib/std/heap/general_purpose_allocator.zig:612]: `AutoHashMap` should be better, especially now that it has a fastpath for easy to hash keys.

**Pull request merged**